File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        stack<int> s;
        ListNode* temp = head;
        while(temp!=NULL){
            s.push(temp ->val);
            temp = temp -> next;
        }

        temp = head;

        while(temp){
            if(s.top() != temp->val) return false;
            s.pop();
            temp=temp->next;
        }
        
        // int j = arr.size()-1;
        // int i = 0;
        // while(i<j){
        //     if(arr[i]!=arr[j])return false;
        //     i++;
        //     j--;
        // }
        return true;
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head){
        ListNode prev = null;
        ListNode curr = head;
        ListNode temp;

        while(curr != null){
            temp =  curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;


        }

        return prev;
    }
    public boolean isPalindrome(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;

        while(fast!=null && fast.next !=null){
            fast=fast.next.next;
            slow = slow.next;

        }

        ListNode second = reverse(slow);

        ListNode first  = head;

        while(second!=null){
            if(second.val != first.val){
                return false;
            }
            
            first=first.next;
            second=second.next;

        }   

        return true;

    }
}